design/0001-metallb-crd.md

```diff
@@ -0,0 +1,112 @@
+interface for MetalLB based on Kubernetes Custom Resource Definitions (CRDs)
+while also retaining support for the existing ConfigMap interface.
+
+## Motivation
+
+Today MetalLB allows only configuration using ConfigMap objects,
+which is simple from configuration perspective and doesn't
+require cluster-admin rights, However it doesn't have validation on the API
+server level, no support for API versioning and no fine-grained permission
+model via RBAC.
+Custom resource definition "CRD" are managed like any other resource,
+through the Kubernetes API, and eventually stored in the backend store Etcd.
+
+### Goals
+
+- Configuring MetalLB using kubernetes custom resource "CRD".
+- Retain support of the existing ConfigMap interface for backwards
+  compatibility.
+
+## Proposal
+
+Using [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+framework to build new CRD "MetalLB" object and a controller which works as
+a translator from the user's configuration yaml to a MetalLB's configmap
+object, with this approach there will be no changes required for the
+MetalLB management plane.
+
+### Risks and Mitigations
+
+Users need to either use CRD or ConfigMap to configure MetalLB
+but not both, if adminstrator mistakenly uses both, the CRD configuration will
+overwrite the user defined ConfigMap.
+
+## Design Details
+
+### Design Overview
+
+[Metallb CRD Design diagram](https://excalidraw.com/#json=5898667890311168,GWkLmmW6WMvZu6JdisIT0Q)
+
+![Metallb CRD Design diagram](./metallb-crd.png)
+
+This proposal will implement three custom resource objects :-
+
+- [AddressPools](https://github.com/metallb/metallb/pull/850/files#diff-e700ddc5833f14f3e89d361ddfc1b1965022f7f15a5fc46dcdba3185f8cafe7b)
+
+- [Peers](https://github.com/metallb/metallb/pull/850/files#diff-e700ddc5833f14f3e89d361ddfc1b1965022f7f15a5fc46dcdba3185f8cafe7b)
+
+- [BgpCommunities](https://github.com/metallb/metallb/pull/850/files#diff-8f9cdf24a9af2238173c872bb20e3a18aa02bbea9908817d29dcdf1283f7fd60)
+
+### Configuration Flow
+
+1. User will supply the yaml configuration, here is a
+   [sample config](https://github.com/metallb/metallb/pull/850/files#diff-18cb180f903e44677f91d9eb946d3cf7230ca48ff708f145722b09031bcfe54a).
+1. Configuration will get authenticated and authorized based on RBAC rules.
+1. Supplied configuration will get verified using openAPIV3Schema based on
+   the definition of CRDs.
+1. New objects will be stored in etcd.
+1. Based on the supplied configuration, All the controllers monitor API server looking for creation of
+   AddressPools/Peers/BgpCommunities resource.
+1. The CRD's controller will use golang based template to convert the configuration object to
+   its crossponding part of the final ConfigMap object.
+1. The generated ConfigMap object will be written to etcd so MetalLB's
+   components can consume it as if the adminstrator directly used ConfigMap.
+
+### CRD Extendability
```

Comment:
This needs more work. Being able to extend CRDs isn't the issue here. We are talking about finding a good data model which can be flexible enough for the currently-known use cases and hopefully some future ones, too.

In other words, it's not enough to say "we can add field X to resource Y", we need to actually figure out if it **makes sense** to do.

For example, consider the following two use cases. They might look identical but they aren't.

1. Announce pool P1 only from node group N1. Do not announce pool P1 from any other node groups.
  - Node group N1 may also announce pool P2.
  - Pool P1 may NOT be announced by node group N2.
2. Announce only pool P1 from node group N1. Do not announce any other pools from group N1.
  - Node group N1 may NOT announce pool P2.
  - Pool P1 may be announced by node group N2.

Given the above, what should we do? Add a node/peer selector to pools? Add a pool selector to peers/nodes? Do both? How do the two knobs interact? Have we thought about side effects?

The above is just a **sample** problem. I'm sure there are more we haven't considered. For example, what about cases where node N peers with routers R1 and R2 and wants to advertise a prefix to R1 but not to R2?

I feel like we need to dedicate more thought to this and discuss such topics in the proposal. As discussed before, **flexibility is a major concern** when touching the config surface. It would be a mistake to add some fields to the CRDs just to get the CRD PR out the door. We need to see that we are _actually_ solving existing problems and aren't introducing new ones. I'm aware we can't foresee everything but we can still think about cases that are already known to us e.g. from open issues.

Reply:
it sounded like you are looking for Node pool announcement ACL conifg with implicit deny something like, assuimg we use
Ids for pool assignments :- 

nodeSelectorGroup: ng1
  node1
  node2

nodeSelectorGroup: ng2
  node3

nodeSelectorGroup ng1:
   Accept: Pool >= Pool1 && Pool < Pool10  // range [1:10]

nodeSelectorGroup ng2:
    Accept:  Pool != Pool1

which another object abd can be extended to Peers too

Reply:
Maybe we could add a commit(s) to this PR to help understand how we would use the CRD and what it would look like. I don't think we will be able to cover every use case but it would at least give us (increasing) confidence on the structure. For example, we could include the following:

1. List of sample uses cases: I don't think each needs to be very detailed (perhaps a single line for each like the example @johananl specified above `Announce pool P1 only from node group N1. Do not announce pool P1 from any other node groups.`). @msherif1234 maybe you can start this? You could use the unit tests for inspiration and then you could get feedback from the rest of the community on them.

2. The proposed CRD or CRDs

3. Examples of how the CRD(s) from (2) could be used to address the use cases in (1). This could show object definitions, for example.

I think if we don't approach it in a structured way, we may miss some corner cases.

Reply:
On a separate point, I think it would be good to outline some kind of plan for how we would version this API? Also, how (and if) it should align to MetalLB releases.

Reply:
@markdgray I have been thinking about nodeselector option today its under Peers conifg I think if needs to move up in the level so it can work for both addresspools and peers I will share what I have once its ready

Reply:
just that it seems more like an attirbute than an independent object to me, and also seems more consistent with other places the selector pattern exists